Build precompiled and publish to S3

wingyplus opened the pull request:
Continuing #12038 from the discussion in https://github.com/elixir-lang/elixir/pull/12053#issuecomment-1208268123. I create a new workflow to publish a precompiled version to s3. Currently, it will run on every commit that pushes to main. If it's too much, we can change it to run every night (with a cron schedule).

wingyplus commented on .github/workflows/publish_precompiled.yml:
```diff
@@ -0,0 +1,52 @@
+          version=$(cat VERSION)
+          for file in elixir-${version}-otp-${{ matrix.otp }}.zip elixir-${version}-otp-${{ matrix.otp }}.zip.sha1sum elixir-${version}-otp-${{ matrix.otp }}.zip.sha256sum
+          do
+            aws s3 cp "${file}" s3://s3.hex.pm/builds/elixir/ \
```
I copied the destination from https://github.com/hexpm/bob/blob/main/priv/scripts/elixir/elixir.sh#L81. But it can be change.

wingyplus replied:
@wojtekmach I'll look into it. :)

I also have a question, do we still need to publish a shasum per file as the release workflow does?

wingyplus commented on .github/workflows/publish_precompiled.yml:
```diff
@@ -0,0 +1,52 @@
+        env:
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          AWS_DEFAULT_REGION: us-east-1
```
What region should use?

wingyplus commented:
These are files that this workflow publishes. The screenshot below is tested on my bucket.

<img width="1174" alt="Screen Shot 2565-09-06 at 21 57 58" src="https://user-images.githubusercontent.com/484530/188669725-c1950f04-49b7-4819-8309-36540f3cc68f.png">

wojtekmach reviewed (approved):
This is great! Besides the one comment I believe all that is left is s3 auth tokens and bumping fastly cache, both of which I can help making happen.

wingyplus commented on .github/workflows/publish_precompiled.yml:
```diff
@@ -0,0 +1,52 @@
+          version=$(cat VERSION)
+          make Precompiled.zip
+          mv Precompiled.zip elixir-${version}-otp-${{ matrix.otp }}.zip
+          shasum -a 1 elixir-${version}-otp-${{ matrix.otp }}.zip > elixir-${version}-otp-${{ matrix.otp }}.zip.sha1sum
+          shasum -a 256 elixir-${version}-otp-${{ matrix.otp }}.zip > elixir-${version}-otp-${{ matrix.otp }}.zip.sha256sum
```
We should change the name to branch like `main-otp-${{ matrix.otp }}` to make it compatible with Bob right?

wingyplus commented on .github/workflows/publish_precompiled.yml:
```diff
@@ -0,0 +1,52 @@
+        run: |
+          version=$(cat VERSION)
+          make Precompiled.zip
+          mv Precompiled.zip elixir-${version}-otp-${{ matrix.otp }}.zip
+          shasum -a 1 elixir-${version}-otp-${{ matrix.otp }}.zip > elixir-${version}-otp-${{ matrix.otp }}.zip.sha1sum
+          shasum -a 256 elixir-${version}-otp-${{ matrix.otp }}.zip > elixir-${version}-otp-${{ matrix.otp }}.zip.sha256sum
```
We should change the name to branch like main-otp-${{ matrix.otp }} to make it compatible with Bob right?

wojtekmach replied:
Yeah, not sure. I think we can rename it to `main-otp-x` just before uploading to Bob.

wingyplus commented:
@wojtekmach I addressed all issues that you commented. Please kindly review again. 🙇

wingyplus commented:
The freebsd raise an error:

```
gmake compile
ld-elf.so.1: /lib/libc.so.7: version FBSD_1.7 required by /usr/local/lib/erlang24/erts-12.3.2.2/bin/beam.smp not found
gmake: *** [Makefile:83: lib/elixir/src/elixir_parser.erl] Error 1
```

I'm not sure why. :(

wojtekmach commented on .github/workflows/publish_precompiled.yml:
```diff
@@ -0,0 +1,70 @@
+            otp_version: 24.3
+          - otp: 25
+            otp_version: 25.0
+    runs-on: ubuntu-18.04
```
18.04 is deprecated:

```suggestion
    runs-on: ubuntu-20.04
```

wingyplus replied:
Fixed

wojtekmach commented on .github/workflows/publish_precompiled.yml:
```diff
@@ -0,0 +1,70 @@
+          date=$(date -u '+%Y-%m-%dT%H:%M:%SZ')
+          build_sha256=$(cut -d ' ' -f 1 main-otp-${{ matrix.otp }}.zip.sha256sum)
+
+          aws s3 cp s3://${{ env.S3_BUCKET }}/builds/elixir/builds.txt builds.txt || true
+          touch builds.txt
```
I'd remove ` || true` and `touch builds.txt` as it could obscure s3 network errors. It seems it is only useful for when there's no builds.txt in s3 already?

wingyplus replied:
Yes it is

wingyplus replied:
I changed to || touch builds.txt instead.:)

wojtekmach replied:
If we got an intermittent network failure on `aws s3 cp` then we would create empty builds.txt whereas I think it is better to crash.

wingyplus replied:
Ahh, I see. I’ll take a look at it

wojtekmach commented on .github/workflows/publish_precompiled.yml:
```diff
@@ -0,0 +1,70 @@
+
+      - name: Upload to S3
+        run: |
+          for file in main-otp-${{ matrix.otp }}.zip main-otp-${{ matrix.otp }}.zip.sha1sum main-otp-${{ matrix.otp }}.zip.sha256sum builds.txt
```
I think you might have already asked about it, do we need to upload shasum files? They are redundant since we have builds.txt so maybe we don't?

wingyplus replied:
I’m not sure if we should keep it like release workflow does?

wojtekmach reviewed (commented):
cc @ericmj

ericmj commented on .github/workflows/publish_precompiled.yml:
```diff
@@ -0,0 +1,69 @@
+          echo -e -n "main-otp-${{ matrix.otp }} ${{ github.sha }} ${date} ${build_sha256}\n$(cat builds.txt)" > builds.txt
+          sort -u -k1,1 -o builds.txt builds.txt
+        env:
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
```
```suggestion
          AWS_ACCESS_KEY_ID: ${{ secrets.HEXPM_ELIXIR_BUILDS_AWS_ACCESS_KEY_ID }}
          AWS_SECRET_ACCESS_KEY: ${{ secrets.HEXPM_ELIXIR_BUILDS_AWS_SECRET_ACCESS_KEY }}
```

ericmj reviewed (commented):
This looks good, we just need to add the Fastly cache purging.

ericmj commented on .github/workflows/publish_precompiled.yml:
```diff
@@ -0,0 +1,86 @@
+          curl \
+            --fail \
+            -X POST \
+            -H "Fastly-Key: ${BOB_FASTLY_KEY}" \
```
```suggestion
            -H "Fastly-Key: ${BOB_FASTLY_PURGE_KEY}" \
```

wingyplus replied:
Fixed. 🙇

wingyplus commented:
@wojtekmach @ericmj I addressed all comment suggestions. And I just added fastly purge cache after build but unfortunately, I've not fastly setup, so I can't test it by myself. :(

wingyplus commented on .github/workflows/publish_precompiled.yml:
```diff
@@ -0,0 +1,86 @@
+            "https://api.fastly.com/service/${BOB_FASTLY_SERVICE}/purge/${BOB_FASTLY_KEY}"
+        env:
+          BOB_FASTLY_KEY: ${{ secrets.BOB_FASTLY_KEY }}
+          BOB_FASTLY_SERVICE: changeme
```
I'm not sure what service name we should use or just use `secrets` instead?

ericmj replied:
Lets use secret here as well.

ericmj commented:
> I addressed all comment suggestions. And I just added fastly purge cache after build but unfortunately, I've not fastly setup, so I can't test it by myself. :(

Thank you. We can test this 👍

ericmj commented on .github/workflows/publish_precompiled.yml:
```diff
@@ -0,0 +1,86 @@
+            -H "Content-Length: 0" \
+            "https://api.fastly.com/service/${BOB_FASTLY_SERVICE}/purge/${BOB_FASTLY_PURGE_KEY}"
+        env:
+          BOB_FASTLY_KEY: ${{ secrets.BOB_FASTLY_KEY }}
+          BOB_FASTLY_SERVICE: changeme
```
```suggestion
          BOB_FASTLY_KEY: ${{ secrets.HEXPM_REPO_FASTLY_KEY }}
          BOB_FASTLY_SERVICE: ${{ secrets.HEXPM_REPO_FASTLY_SERVICE }}
```

wingyplus replied:
@ericmj Changed. And rename BOB_FASTLY_* to just FASTLY_*. :)

paulo-ferraz-oliveira commented:
Will this eventually mean we replace the place `setup-beam` fetches the pre-compiled from (https://github.com/erlef/setup-beam/blob/main/src/install-elixir.sh#L12)? Or is this just a normal redirect to S3 (thus no changes to `setup-beam`)?

wingyplus commented:
> Will this eventually mean we replace the place `setup-beam` fetches the pre-compiled from (https://github.com/erlef/setup-beam/blob/main/src/install-elixir.sh#L12)? Or is this just a normal redirect to S3 (thus no changes to `setup-beam`)?

I’m expecting that there’s no changes on someone who fetch pre-compiled version. The file still put to the same place like before. :)

wojtekmach commented on .github/workflows/publish_precompiled.yml:
```diff
@@ -0,0 +1,86 @@
+      - name: Build Elixir Release
+        run: |
+          make Precompiled.zip
+          mv Precompiled.zip main-otp-${{ matrix.otp }}.zip
```
repo.hex.pm also has main.zip (without -otp-X) which is compiled against oldest supported OTP. We should support it as well and then I think we can ship it.

wingyplus replied:
@wojtekmach So in this case it should build with otp24 right?

wojtekmach replied:
Yes

wingyplus replied:
Let me try

wingyplus replied:
@wojtekmach main.zip added in https://github.com/elixir-lang/elixir/pull/12118/commits/6a2c7d5dd75fdd3d93ee7734920c531dc313601f. Please kindly review.

wingyplus commented:
Hi,

i just see the issue https://github.com/ScoopInstaller/Main/issues/3886#issuecomment-1239843242. I start thinking that we should bring back the shasum file per version for consumer usage reason. And it will consistent with current release workflow.

what do you think?

wojtekmach commented:
Per https://github.com/elixir-lang/elixir/issues/12134 scoop might go with repo.hex.pm/builds/elixir/* and perhaps theyd use builds.txt for checksums

wingyplus commented:
> Looks great. We need to coordinate merging this with changing bob.

I've open a PR https://github.com/hexpm/bob/pull/125 to remove main. But it seems we need to have building doc feature before removing it. If have anything need to do more, feel free to comment. 🙇

wojtekmach commented:
I believe docs is the remaining thing!

wingyplus commented:
Yeah, i’ll try tackle that after this changes get merge.

wojtekmach commented:
I believe we can only merge this after we stop Bob from building elixir main as not to keep overwriting builds.txt. But we can only stop bob from building main if we can build docs on GitHub. So I think we should do it as part of this PR. WDYT?

wingyplus commented:
> I believe we can only merge this after we stop Bob from building elixir main as not to keep overwriting builds.txt. But we can only stop bob from building main if we can build docs on GitHub. So I think we should do it as part of this PR. WDYT?

I’m ok to continue on this PR.
